### ISSUE TITLE
One program, multi modules to their respective files.

### ISSUE THREAD

**Issue by 92JBach** [Delgan/loguru#54]
I am fairly new to logging, I used print statements everywhere just to debug.

Now I'm trying to implement a logger for my program I plan to  pass it to each sub-module as the program is running and I want it to have each module output their logs to their own file.

I know this is a simple question and will probably take little to no time to answer I just need clarification.

For context it's a Discord bot, and in my `launcher.py` file I'm looking to add a logger instance and just pass it to each cog where they can use it.

If this is not good practice, and I should have each module use their own logger then that's fine it just seemed best to do this way.

Also, as I'm new to logging what is the use of multiple `logger.add(...)`'s? Is this to format how `logger.INFO`, `logger.DEBUG`, `logger.ERROR` output? As in if I use it multiple times to multiple files I would in essence be re-establishing where it outputs. Example:

    logger.add("file1.log", format="{time} {level} {message}", filter="my_module", level="INFO")
    logger.add("file2.log", format="{time} {level} {message}", filter="my_module", level="INFO")

Would these output the same message to each file?

**Comment by Delgan** [comment]
Hi @92JBach. 

Each time you call [`.add()`](https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.add), you create a new handler to which each logged message will be propagated. An handler is used to define what to do with the logged message.
If you `.add("file1.log")` then `.add("file2.log")` and finally call `logger.debug("Some message")`, the two files will both contain the logged message.

The main use case for multiple handlers is to have one which logs to `sys.stderr` (in live in your terminal, as do the default handler) and another one which logs to `"some_file.log"` (to save on disk what happened). 

For example, you could also add a third handler like `logger.add(function_to_send_an_email, level="CRITICAL")`. Then, whenever you call `logger.critical("Error!")` from anywhere in your program because something gone wrong, you will receive an e-mail warning you that you program crashed.

This is why multiple handlers are useful. You can do whatever you want with them.

In particular, and that's of your interest here, you can add one file-handler per sub-module. I would say it's usually better if sub-modules are logged to the same file, as it allows you to see your code flow as a whole. However, logging to one file per module is perfectly doable if you prefer!

I think calling `logger.add("sub_module_1.log", filter="sub_module_1")` should do the trick, thanks to the `filter` attribute. This means that the `"sub_module_1.log"` file will only contain messages that you logged within `sub_module_1`.
Also, I don't know how you intend to "pass" the logger to your sub-modules, but you should probably just do `from loguru import logger` . They will send logs to the added handlers anyway.

Don't hesitate to ask further questions or clarifications. 😉

**Comment by 92JBach** [comment]
Thank you for the patient guidance. I feel like such a noob asking these things *and I'd consider myself pretty fluent in python with the exception of me overthing things :laughing:*

*just so you know, I did read the documentation. I'm just again new coupled with some of the advanced implementations confusing me. I just need to get the simpler things down and I should have no trouble figuring out the rest.*

As for clarification, doing `from loguru import logger` in each submodule will remember each added handler? As in no need to pass logger? It just seemed too good to be true! So in what scope does it `drop` *(close/remove)* them? Upon application termination so each app is bounded to their own "logger"?

As for `filter` what exactly can I use for this and in what ways is it called by default. Ex:

**in a sub-module**

    @logger.catch()
    def foo:
        try:
            #some risky business
        except Exception as e:
            #Send end user an "oops I can't do that" message (for example)
            raise e

So say the sample above, would this be correct useage and wondering if I can modify or override catch and implement a different message be sent to the user for each different error instead of doing a try except block in each function I'd use `catch`

**Comment by Delgan** [comment]
Yeah, don't worry, each time I have to learn a new library I have a bunch of questions too and I need some time before starting to grasp it correctly. 🙂 

> As for clarification, doing `from loguru import logger` in each submodule will remember each added handler? As in no need to pass logger?

Yes, exactly. The `logger` is just an object which dispatch your logs to all added handlers. Usually, you should `.add()` handlers in your main module while starting your application. Then in your sub-modules, `from loguru import logger` suffices, each future calls to logging functions (`.debug()`, `.info()`, etc.) will be propagated to the configured handlers.

> So in what scope does it `drop` _(close/remove)_ them? Upon application termination so each app is bounded to their own "logger"?

Yes, exactly, again. When your applications ends, the handlers are closed, and that's it. If you run your application a second time, the handlers will be restarted. For example, if you used `logger.add("file.log")` this will continue to write logs at the end of your file, but if you used `logger.add("file_{time}.log")`, this will create a new file each time you run your application.
It is also possible to manually stop the handlers before the program termination by using the [`.remove()`](https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.remove) method.

> As for `filter` what exactly can I use for this and in what ways is it called by default.

The `filter` argument is just a function in charge of deciding for each message if it should or not be passed to the corresponding handler. By default, all incomming logs are passed to the logger. For convenience, the `filter` function can also be a `str`, in this case, only logs from the module are accepted.

> **in a fictional sub-module**
> 
> ```
> @logger.catch()
> def foo:
>     try:
>         #some risky business
>     except Exception as e:
>         #Send end user an "oops I can't do that" message (for example)
>         raise e
> ```
> 
> So say the sample above, would this be correct useage? Also wondering if I could modify or override `logger.catch` in a way as to format messages be sent to the user dependent on the error instead of doing a try except block in each function I'd use `catch`.

I don't understand what your code sample as to do with your question about the `filter` argument?
The [`logger.catch()`](https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.catch) is a convenient decorator used to avoid the `try / except` block, as it implements it internally. So, in this case, it is not very useful to repeat `@logger.catch()` and `try: ... except: ...`.

Instead, you can do:

```python
@logger.catch(reraise=True, message="Oops, I can't do that")
def foo():
    # Some risky business
```

This is equivalent to:

```python
def foo():
    try:
        # Some risky business
    except Exception as e:
        logger.exception("Oops, I can't do that")
        raise e
```


> Would it be bad practice *(or a good case here to make my own customizable level called maybe "NOTIFY" and filter the ones I want to goto that and use `logger.add(send_user_a_message, level="NOTIFY")`
> 
> tricky part for me to wrap my noggin around is how do I filter multiple cases for this? Such as `ValueError`, `IndexError`, `etc`... Let me know If I'm just making more work for myself or if I'm running the wrong way here lol.

Why do you want to filter your logs based on the exception type? What do you want to do if this is a `ValueError` compared to an `IndexError`?

The `level` is not used to filtering, it just indicates the severity of log messages you want to handle.
For this kind of "multiple cases logging", the recommended solution is to use [`.bind()`](https://loguru.readthedocs.io/en/stable/api/logger.html#loguru._logger.Logger.bind) combined with `filter` argument.

```python
logger.add(first_function, filter=lambda record: "first" in record["extra"])
logger.add(second_function, filter=lambda record: "second" in record["extra"])

# This add "first: True" to the "extra" dict of each logged message using this logger
first_logger = logger.bind(first=True)  
# Same for "second: True", but the extra dict doesn't contain "first: True"
second_logger = logger.bind(second=True)

# This is logged to the "first_function", but filtered-out for the "second function"
first_logger.info("Message 1")
# Same for "second_function"
second_logger.info("Message 2")

```
You can also do the dispacthing in your custom handler function, based on the [`record`](https://loguru.readthedocs.io/en/stable/api/logger.html#record):

```python
def send_user_a_message(log_message):
    record = log_message.record
    exception = record["exception"]
    if exception is None:
        print("No error")
    elif exception.type == ValueError:
         print("Value Error")
    elif exception.type == IndexError:
        print("Index error")
```

However, I'm wondering about your use case here.


> Thanks again for the speedy and patient assistance _(hopefully someone else finds these questions helpful)_

No problem, I hope I can help you to use Loguru with ease. 👍

**Comment by 92JBach** [comment]
Thanks buddy, this answers my questions! I think with my understanding the basics now I will be able to figure out the rest. Thanks for making things clear and including the great examples for me!

I do have one more question for you though, is there a way you can fix the [docs](https://loguru.readthedocs.io/en/stable/api.html) so they have tabs for each section? As shown below:

![image](https://user-images.githubusercontent.com/32350542/52527520-d531b800-2c97-11e9-8282-eb7ebdd6240a.png)

There isn't a way to click through each section it just shows `loguru.logger`. My thinking is there should be drop downs or sections for `handler`, `sink`, .. Also, this means there's no ¶ hyperlink at the end of each method. Kind of like how Python docs are.

This is really just a suggestion nothing major.

**Comment by Delgan** [comment]
> Thanks buddy, this answers my questions! I think with my understanding the basics now I will be able to figure out the rest. Thanks for making things clear and including the great examples for me!

Great! In doubt, just @ping me!



> This is really just a suggestion nothing major.

This is a good suggestion, thanks!
I tried to do this in the past, but I had many problems, `Sphinx` (the library used to generate the documentation) doesn't seem suited for this.
Instead of adding this to the sidebar, I think I could try to add it to the body (which is quite useless for now):
![1549784539-screenshot](https://user-images.githubusercontent.com/4193924/52531110-b1f71f00-2d10-11e9-8f01-b9e43442893f.png)

**Comment by 92JBach** [comment]
Haha aren’t web interfaces a hassle sometimes. Thanks again buddy!

**Comment by Delgan** [comment]
As suggested, I populated [the API reference](https://loguru.readthedocs.io/en/latest/api.html) with links to methods and sub-sections.

This looks quite ugly, but it should do the job. 😁

**Comment by jinfagang** [comment]
@Delgan Hi, I have a huge project, using loguru as logging lib, but I have multiple modules, while only one single main.py

I using `logger.add` to add a file handler in main, but when import loguru in child modules, and logger,info, nothing saved int file.
I want this file contains all logs in all modules.

Why?

**Comment by Delgan** [comment]
@jinfagang It's hard to tell, I'm sorry. Can you open a new issue with some minimal reproducible example please?

**Comment by durgeksh** [comment]
I want to have two loggers.
1. Only display to console
2. Display to console and save in file

How this can be achieved?
Thank you for wonderful library.

**Comment by durgeksh** [comment]
> @jinfagang It's hard to tell, I'm sorry. Can you open a new issue with some minimal reproducible example please?

I want to have two loggers.
1. Only display to console
2. Display to console and save in file

How this can be achieved?
Thank you for wonderful library.

**Comment by Delgan** [comment]
@durgeksh Please take a look at these examples in the documentation: [Creating independent loggers with separate set of handlers](https://loguru.readthedocs.io/en/stable/resources/recipes.html#creating-independent-loggers-with-separate-set-of-handlers).

If you need further assistance, please open a new issue so we can discuss it in details in a appropriate ticket.
